src/dfu/dfu_main.cpp

```diff
@@ -0,0 +1,1002 @@
+        cerr << "Error: Value for transport block is not an integer. Given value: " << block_number_str << endl;
+        exit(HOST_APP_ERROR);
+    }
+    if (stoi(block_number_str) >= INVALID_TRANSPORT_BLOCK_NUM)
+    {
+        cerr << "Error: Max value for transport block is " << INVALID_TRANSPORT_BLOCK_NUM-1 << ". Given value: " << block_number_str << endl;
+        exit(HOST_APP_ERROR);
+    }
+    remove_opt(argc, argv, index, 2);
+
+    return stoi(block_number_str);
+}
+int main(int argc, char ** argv)
+{
+    if(argc == 1)
+    {
+        cout << "Use --help to get the list of options for this application." << endl
+        << "Or use --list-commands to print the list of commands and their info." << endl;
+        return 0;
+    }
+    YAML::Node config;
+    int* i2c_info = new int[1];
+    int* spi_info = new int[2];
+    string yaml_file_name;
+
+    // Check optional arguments
+    check_verbose_mode(&argc, argv);
+    uint16_t start_block_number = check_upload_start(&argc, argv);
+
+    // Load YAML file with transport settings
+    yaml_file_name = get_executable_path() + "/transport_config.yaml";
+    if (verbose_mode) {
+        cout << "Parsing YAML file " << yaml_file_name << endl;
+    }
+    if (is_file_found(yaml_file_name)) {
+        config = YAML::LoadFile(yaml_file_name);
+        // Read I2C parameters from YAML file
+        i2c_info[0] = config["I2C_ADDRESS"].as<int>();
+        spi_info[0] = config["SPI_MODE"].as<int>();
+        spi_info[1] = 1024;
+    } else {
+        cerr << "Error: File \'" << yaml_file_name << "\' not found" << endl;
+        exit(HOST_APP_ERROR);
+    }
+
+    // Load dynamic library with transport drivers
+    string device_dl_name = get_device_lib_name(&argc, argv, options, num_options);
+    int * device_init_info = NULL;
+    if(device_dl_name == device_i2c_dl_name)
+    {
+        device_init_info = i2c_info;
+    } else if(device_dl_name == device_spi_dl_name)
+    {
+        device_init_info = spi_info;
+    }
+    string device_dl_path = get_dynamic_lib_path(device_dl_name);
+    dl_handle_t device_handle = get_dynamic_lib(device_dl_path);
+
+    device_fptr make_dev = get_device_fptr(device_handle);
+    Device * device = make_dev(device_init_info);
+
+    control_ret_t ret = device->device_init();
+    if (ret != CONTROL_SUCCESS)
+    {
+        cerr << "Could not connect to the device: error " << ret << endl;
+        exit(HOST_APP_ERROR);
+    }
+
+    // Check first CLI options which don't require access to the device
+    const opt_t * opt = nullptr;
+    int cmd_indx = 1;
+    string next_cmd = argv[cmd_indx];
+    if(next_cmd[0] == '-')
+    {
+        opt = option_lookup(next_cmd, options, num_options);
+        if (opt->long_name == "--help")
+        {
+            return print_help_menu();
+        }
+        else if (opt->long_name == "--app-version")
+        {
+            cout << current_host_app_version << endl;
+            return 0;
+        }
+    }
```

Comment:
It is a matter of an app design, but this app will require a device connection even if the user wants to do --help. xvf_host executes all non-device related options before it connects to the device so that the user can execute some options without the hardware. I'm not saying you should do the same, just keep it in mind, it might be a bit nicer.

Reply:
This should be fixed now,